apps/builder/app/shared/webstudio-data-migrator.ts

```diff
@@ -0,0 +1,43 @@
+import { camelCase } from "change-case";
+import {
+  getStyleDeclKey,
+  type StyleDecl,
+  type WebstudioData,
+} from "@webstudio-is/sdk";
+import { toValue, type StyleProperty } from "@webstudio-is/css-engine";
+import { expandShorthands, parseCssValue } from "@webstudio-is/css-data";
+
+/**
+ *
+ * Transform loaded data and sync into server
+ * before passing into application
+ * should be idempotent because can be executed multiple times
+ *
+ * Very basic version of client migrations
+ * Should be versioned eventually to avoid running every time
+ *
+ * For now patch is prevented by excluding empty transactions from sync queue
+ *
+ */
+export const migrateWebstudioDataMutable = (data: WebstudioData) => {
```

Comment:
1. migrations should be a separate folder with a file per migration type of structure imo
2. we should avoid having to run them all all the time, there should be some way to say that a mgiration isn't needed without iterating over data. Imaging having dozens or hundreds of them

Reply:
You would probably want to have a flag in data for each migration and once there is a flag, e.g. migrations.overflowShorthand === true then its not executing next time

Reply:
Yes, in the future we will have it all. For now the we iterate once over all styles and when nothing is mutated just do not send patch.

Whole migrations stuff would take time to implemented. Here's very fast solution. It does the job.

Reply:
I think adding flags is relatively easy to do and also better organizing migrations on a file system, its really just a flag

Reply:
just a flag will require migration in the future to support dozens
we also need to support sdk dependencies
so not just a flag

Reply:
Why not set the flag as part of json data? so it can be sent along with patches and no migration needed?

Reply:
You are basically making this migration functionality for everyone to use, so it should be good enough, otherwise its just yet another hack added to the system

Reply:
I wouldn't feel comfortable using it knowing it can lead to execution every time. I would also like to have a nicer interface for migrations

Reply:
Some things aren't meant to be incrementally added iwh some naive version first

Reply:
I understand we are under pressure with the release, but that can be postponed if anything

Reply:
I don't know if this is part of our requirement for the current release

Reply:
This flag you want to add will have to be set in data forever. And then we will bring another solution for migrations which will also be set in data forever. So we will have to support both solutions instead of just one which will emerge from what I implemented.